Update Components, fix lint errors, fix bugs and other improvements

michael-r-elp opened the pull request:
With the initial migration to svelte 5 now done, this PR aims to migrate and improve some components as well as do a general look at the whole codebase for potential eslint errors that require looking into.

## Current Notable changes:
### Improvements
* Use BeatSaverClientFactory in search bar
* Use await block in search bar, allows for displaying a message depending on the promise state
* Adds prettierignore as second ignore source for eslint
* Better search preview card, now includes diffs design might close #98 
* Added contributions, closes #140 
* Added ErrorHandler component to catch components that might break the whole page (shows error component instead)

### Components Updates
* Updated the following components to svelte 5 syntax
  - Navbar, Search, MapCards, MapPreview, OneClickDownloadButton, PlaylistCards, PostPage
* Improved Search component

### Fixed Lint errors
* Added keys to all "each" blocks see [require-each-key](https://sveltejs.github.io/eslint-plugin-svelte/rules/require-each-key/) and [keyed-each-blocks](https://svelte.dev/docs/svelte/each#Keyed-each-blocks) for more information 
(technically we don't use anything that requires this I believe, but eslint complains about it so I just add keys)

### Fixed bugs
* Closes #141 

Also I think with us using svelte 5 now, this opens up the possibility to also improve the way certain parts of the website work, but that might be beyond the scope of this PR.



Discussion as well as suggestions about any changes made would be appreciated

TimothyGillespie commented on src/lib/PaginationBar.svelte:
```diff
@@ -20,7 +20,7 @@
     <Fa icon={faAngleLeft} />
   </a>
-  {#each pageNumbers as singlePageNumber}
+  {#each pageNumbers as singlePageNumber (singlePageNumber)}
```
Perhaps the current page number with a separate symbol would be more reliable in cases of SPA like navigation. Not sure, though.

TimothyGillespie commented on src/lib/Search.svelte:
```diff
@@ -3,7 +3,35 @@
+      downVotes: number
+      avgScore: number
+    }
+  }
```
NIce! Feel free organize the typing files differently as well since they became one large monolith.

michael-r-elp replied:
This will probably have to be done in a separate PR by anyone who has some time and a rough idea on how to organize everything

michael-r-elp replied:
This will probably have to be done in a separate PR by anyone who has some time and a rough idea on how to organize everything

TimothyGillespie commented on src/lib/Search.svelte:
```diff
@@ -26,35 +54,27 @@
-  let searchUrl: string
+  let searchUrl: string = $state('')
+
+  let beatSaverPromise: Promise<any> | undefined = $state()
```
The type could possible be derived from the function. But doesn't have to be. Just an idea.

michael-r-elp replied:
The type is the Preview type I have at the top, just forgot to replace this

TimothyGillespie commented on src/routes/posts/search/+page.svelte:
```diff
@@ -37,7 +37,7 @@
 <h1>Article Search</h1>
 <div class="cards">
-  {#each posts as post}
+  {#each posts as post (post)}
```
This doesn't seem right. Does this work to pass in a whole object in svelte 5?

michael-r-elp replied:
Yes apparently it does, though not recommended according to the docs https://svelte.dev/docs/svelte/each#Keyed-each-blocks

> If a key expression is provided — which must uniquely identify each list item — Svelte will use it to diff the list when data changes, rather than adding or removing items at the end. The key can be any object, but strings and numbers are recommended since they allow identity to persist when the objects themselves change.

michael-r-elp commented on src/lib/Leaderboard.svelte:
```diff
@@ -22,7 +22,8 @@
 <Listing {title} {img} {linkText} {linkUrl}>
-  {#each players as player, index}
+  <!-- Using player.name as key but ideally it should be some kind of uid -->
+  {#each players as player (player.name)}
```
Well I have found out that while they recommend to use a string or number as key, it's also possible to just use the object itself, so I might change it to that in places like here

michael-r-elp commented:
Changed the buttons on the map/playlist cards, basically if a device doesn't support hover for example a tablet we always display the buttons if the available space on the screen allows for it.

What I'm not sure about right now is whether the oneclick download buttons should be displayed on non-hover capable devices or not. I can't really imagine a touch screen device where you'd use one click downloads on, except maybe one of those windows tablets, also I'm unsure on the behaviour if a device has both a mouse and touchscreen, I believe it would be considered hover capable maybe? Could probably test by plugging in a mouse into a smartphone and tablet and see how it reacts.

michael-r-elp commented:
I'll mark this ready for review now, as there are quite a lot of changes in here now.

And I unfortunately lack the time to do more in-depth work on this, so I rather get what we have done and merged and someone else or if I happen to have time again will have to go through everything and make improvements as well as update remaining components to svelte 5 syntax where necessary.

TheCzar1994 commented:
I will try to take a look at this as well at some point this week - hopefully tomorrow

TheCzar1994 commented:
Looks like the Featured Playlists page is broken. I'm just getting a gray screen.
https://deploy-preview-222--bsaber.netlify.app/playlists/page/1

michael-r-elp commented:
> Looks like the Featured Playlists page is broken. I'm just getting a gray screen. https://deploy-preview-222--bsaber.netlify.app/playlists/page/1
> 
> I haven't looked at any code yet but wanted to note other things I find here for when I do:
> 
> * Featured Playlist cards on homepage don't have download buttons on mobile/touch screens where MapCards and MOTW do. This seems inconsistent.
> * MapCard song preview hover seems to either duplicate the background color causing a border or there is some border issue or perhaps before/after issue?

That should all be fixed now.

The page broke again because I thought I could remove the workaround if I have the elements nested inside another div, but turns out that only works for the front page playlists somehow.

Added the download buttons to the playlists for mobile, though they seem quite small so idk how usable they actually end up beeing, at least it's consistent now.

For the MapCard Song Audio Preview, I think I changed it to a button element at some point and didn't properly account for any default styles that come with it, I fixed that now, disabled the border and box shadow style that the browser adds.

michael-r-elp commented:
For additional safety where a component erroring might break the whole site, I added a handler component and used a svelte:boundary for the playlistscards in the playlists page this can be useful in case the problem occours again, while also showing some helpful information for debugging.

Usage:
```svelte
<script>
  import ErrorHandler from '$lib/ErrorHandler.svelte'
</script>

<svelte:boundary>
  <FlakyComponent />

  {#snippet failed(error, reset)}
    <ErrorHandler
      error={error}
      reset={reset}
    />
  {/snippet}
</svelte:boundary>
```

See above image example of the ErrorHandler.svelte Component

![image](https://github.com/user-attachments/assets/1dc60926-ddd2-4a44-b1fc-0460bd83dc7c)


I don't expect people to really come accross this, but it's there now in case we run into a problem again.

TheCzar1994 commented on src/lib/MapOfTheWeekSection.svelte:
```diff
@@ -119,9 +119,9 @@
               </button>
             {/if}
             <div class="interactive-buttons">
-              <CopyBsr mapId={mapOfTheWeek.map.id} />
```
I'm confused why this change is here? Swapping these around and adding the `direction: rtl` to the styling actually causes an uneven spacing change to these buttons
With rtl (uneven spacing)
![image](https://github.com/user-attachments/assets/e3e1d400-fe0a-4264-a6f4-74534e184e3c)
Without rtl (even spacing)
![image](https://github.com/user-attachments/assets/3b0332ec-c972-42c7-b143-265afe1f53c1)

michael-r-elp replied:
Because on touch devices some buttons are hidden, leaving an empty space where the button would be, this was one possible solution for it.

michael-r-elp replied:
The uneven spacing seems to be because the actual svg within the a element is not centered.
I'll see about fixing it later if I remember.

TheCzar1994 replied:
This looks fine, thanks! Sorry it's taken so long to get this reviewed. Will try my best to get it done this week!! Or hopefully even still tonight!

TheCzar1994 commented on src/lib/Search.svelte:
```diff
@@ -135,15 +179,25 @@
       }
     }, 400)
   }
+
+  // Ideally I'd do the below using CSS but
+  // the svelte compiler seems to remove these css classes
+
+  // Function to calculate color based on percentage takes percentage as float
+  function getGradientColor(rating: float) {
+    // Map rating percentage (0-100) to hue range (0 = red, 120 = green)
+    const hue = rating * 120 // 0% = red, 100% = green
+    return `hsl(${hue}, 100%, 50%)` // Full saturation and 50% lightness
+  }
```
I've got to be honest; I'm not a huge fan of how this looks. There were complaints previously about the color coding of ratings on BeatSaver and I'd prefer to avoid that drama. However, I'm also unsure of how else to represent it so not sure how to go about this. Perhaps just a flat color? We've already got the color coded CaretUp and CaretDown

TheCzar1994 replied:
To be fair, the initial complaints on the BeatSaver side were with the pill style rating bar below the album covers which are now orange and green. It used to be red and green - and when there were more downvotes than upvotes, the red would overtake the space and would cause morale issues for some people. I suppose that is different than this and is unlikely to cause anywhere near as much of an issue.

I guess at the end of the day, I'm not entirely opposed to this and I also don't have another answer to it either, so I'm fine with merging this as is and perhaps we can look at it again later if needed.

TheCzar1994 commented on src/routes/community-hub/+page.svelte:
```diff
@@ -2,18 +2,18 @@
-  } from '@fortawesome/free-brands-svg-icons'
-  import { faLink, faBook } from '@fortawesome/free-solid-svg-icons'
-  import Fa from 'svelte-fa'
+  // import {
+  //   faDiscord,
+  //   faFacebook,
+  //   faInstagram,
+  //   faTwitch,
+  //   faXTwitter,
+  //   faYoutube,
+  //   faReddit,
+  //   faGithub,
+  // } from '@fortawesome/free-brands-svg-icons'
+  // import { faLink, faBook } from '@fortawesome/free-solid-svg-icons'
+  // import Fa from 'svelte-fa'
```
Looks like these can be removed instead of commented out?

TheCzar1994 commented on src/routes/playlists/eminem-pack/AlbumPlaylistCard.svelte:
```diff
@@ -1,7 +1,7 @@
   import MapCards from '$lib/MapCards.svelte'
   import OneClickDownloadButton from '$lib/OneClickDownloadButton.svelte'
-  import SocialIcon from '$lib/SocialIcon.svelte'
+  // import SocialIcon from '$lib/SocialIcon.svelte'
```
Can remove this if not needed

TheCzar1994 commented on src/routes/playlists/love-live-birthday-pack/AlbumPlaylistCard.svelte:
```diff
@@ -6,7 +6,7 @@
   export let playlistId: number
 
-  const playlistUrl = `https://api.beatsaver.com/playlists/id/${playlistId}/download/beatsaver-${playlistId}.bplist`
+  // const playlistUrl = `https://api.beatsaver.com/playlists/id/${playlistId}/download/beatsaver-${playlistId}.bplist`
```
Can remove if not needed

TheCzar1994 reviewed (commented):
Overall, this looks good to me. I especially like the revisions to the search bar so thank you for your work on that! I left a few mostly minor comments. I would still like to review this more on mobile/touch devices and perhaps on my Quest, though, with BMBF being deprecated I'm unsure how much Quest users need to be able to have the buttons anyways.

Looks like there's still some open comments from @TimothyGillespie too?

michael-r-elp commented on src/lib/Difficulties.svelte:
```diff
@@ -45,6 +45,7 @@
     align-items: center;
     font-size: 0.75rem;
     user-select: none;
+    // container-type: inline-size;
```
It was added and then commented out, there's some issue with the style of the difficulties at certain breakpoints since they're intended for use in map cards that at certain sizes are next to each other.

michael-r-elp commented on src/lib/Difficulties.svelte:
```diff
@@ -154,6 +155,10 @@
     display: none;
   }
 
+  // Switch to the below whenever there's proper support for it in svelte/sass
+  // @container (width > 650px) {
+
+  // Find a solution for the below that also properly works for search cards
```
Yeah relates to my above comment about sizing in map cards unfortunately `@container` didn't seem to be supported in sass.

TheCzar1994 commented on src/lib/MapCards.svelte:
```diff
@@ -205,22 +241,50 @@
     );
-    background-size: $background-size;
+
+    &:hover {
+      background: linear-gradient(
+        90deg,
+        $color-background-primary $gradient-start,
+        $color-background-tertiary 100%
+      );
+    }
```
Sorry, hopefully last few comments before I'm good with this...
It looks like this is causing a duplication of the effect on desktop. I deleted the:
```
   background: linear-gradient(
      90deg,
      $color-background-primary $gradient-start,
      $color-background-secondary 100%
    );
```

above this but I'm guessing that will cause issues with it on touch/mobile since you/we want it to have that normally there? I honestly don't know how to test it on touch devices so not sure what the best fix would be.

michael-r-elp replied:
Can you go a bit more into detail what you mean with duplication, it looks basically the same for me, except that I missed the `background-size: 160%` on hover for desktop, which makes the border slightly less prominent when hovering over, but otherwise it looks basically the same as on the production site to me

## Desktop:
### This PR with hover and not hover, plus the change I mentioned:
![image](https://github.com/user-attachments/assets/3b4fee34-ba28-4284-9ea8-d0cc5b151a8a)

### Production Page:
![image](https://github.com/user-attachments/assets/5097d265-4a8c-4cf7-92ec-d60deb4ed47e)


## Mobile:
### This PR:
![image](https://github.com/user-attachments/assets/c54e0508-ea6d-4531-a603-25a6631c6e49)

### Production Page:
![image](https://github.com/user-attachments/assets/7d246172-0bcd-48c4-8695-3a4a4ba5a507)

TheCzar1994 commented on src/lib/PlaylistCards.svelte:
```diff
@@ -85,16 +88,31 @@
   }
 
   a .zip-download-button-container {
-    right: 2.5rem;
-    padding: 0.3em 0.45em;
+    right: 0.3rem;
+    padding: 0.3rem;
```
Again, hard to test without knowing a good way to test on touch devices but I think this change causes the button to not be a circle on touch devices? 
![IMG_0213](https://github.com/user-attachments/assets/0fc6da3e-0bcc-4a2a-aae5-8b21d2b8fd1b)

michael-r-elp replied:
Oh yeah, that one was my mistake should be `padding: .3em .45em;` I'll fix that
